py: convert None/False/True to small immediate values

dpgeorge opened the pull request:
Based on discussion in #5314, with full credit to @jongy for the idea:

> mp_const_none is widely used throughout the project 
> ...
> In order to make the loading of it easier, we can either make the value a small immediate like MP_OBJ_NULL

That's what's done in this PR, to make `None`, `False` and `True` small immediate values, namely `MP_OBJ_NONE`, `MP_OBJ_FALSE`, `MP_OBJ_TRUE`.

I had to steal some of the qstr encoding space in an object (see change to py/mpconfig.h) to encode these new values, so they were separate from true object pointers.  But it turned out pretty neat and straightforward to do it.  Hardly any places in the code assume anything about the None/False/True objects so they can easily become immediate values.

There are significant size reductions to ports:
```
   bare-arm:  -416 -0.628% 
minimal x86:  -308 -0.207% [incl -12(data)]
   unix x64:  -992 -0.199% [incl -256(data)]
      stm32: -1928 -0.507% PYBV10
     cc3200: -1240 -0.670% 
      esp32:  -528 -0.047% GENERIC[incl -32(data)]
        nrf:  -932 -0.638% pca10040
       samd:  -560 -0.549% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Note that I only made it work for object representation A at this point.  Also, did not do any performance comparison tests with the changes.

dpgeorge committed:
py/nativeglue: Use mp_const_X instead of &mp_const_X_obj.

Jongy commented:
Very cool! I had the idea earlier of increasing the number of bits used for the object tag in order to add a new tag for those objects, but I thought it was too far to go. Reusing the qstr tags for the price of reducing the bits for qstrs (which anyway don't make any use of the full bits) is a good trade-off.

Jongy commented:
Btw, I wrote a short blog post about the idea behind #5320, I also mentioned this new PR there (see it [here](https://medium.com/@yon.goldschmidt/low-level-code-size-optimizations-in-c-5b9e17d288ce))

dpgeorge commented:
>  I had the idea earlier of increasing the number of bits used for the object tag in order to add a new tag for those objects

I also thought about this but it'd mean aligning all static/ROM objects to 8 bytes (so the 3 lower bits of the pointer are 0) which would likely lead to code size increases.

dpgeorge committed:
py/obj.h: Redefine qstr object encoding to add immediate obj encoding.

This commit adjusts the definition of qstr encoding in all object
representations by taking a single bit from the qstr space and using it to
distinguish between qstrs and a new kind of literal object: immediate
objects.  In other words, the qstr space is divided in two pieces, one half
for qstrs and the other half for immediate objects.

There is still enough room for qstr values (29 bits in representation A on
a 32-bit architecture, and 19 bits in representation C) and the new
immediate objects can be used for things like None, False and True.

dpgeorge committed:
py/obj: Add MICROPY_OBJ_IMMEDIATE_OBJS option to reduce code size.

This option (enabled by default for object representation A, B, C) makes
None/False/True objects immediate objects, ie they are no longer a concrete
object in ROM but are rather just values, eg None=0x6 for representation A.

Doing this saves a considerable amount of code size, due to these objects
being widely used:

   bare-arm:  -392 -0.591%
minimal x86:  -252 -0.170% [incl +52(data)]
   unix x64:  -624 -0.125% [incl -128(data)]
unix nanbox:    +0 +0.000%
      stm32: -1940 -0.510% PYBV10
     cc3200: -1216 -0.659%
    esp8266:  -404 -0.062% GENERIC
      esp32:  -732 -0.064% GENERIC[incl +48(data)]
        nrf:  -988 -0.675% pca10040
       samd:  -564 -0.556% ADAFRUIT_ITSYBITSY_M4_EXPRESS

Thanks go to @Jongy aka Yonatan Goldschmidt for the idea.

dpgeorge committed:
py/obj: Optimise mp_obj_get_type for immediate objs with repr A and C.

This function is called often and with immediate objects enabled it has
more cases, so optimise it for speed.  With this optimisation the runtime
is now slightly faster with immediate objects enabled than with them
disabled.

dpgeorge commented:
This has now been merged.

Thanks @Jongy for the great idea!
